genai-perf/genai_perf/export_data/telemetry_data_exporter_util.py

```diff
@@ -0,0 +1,190 @@
+# Copyright 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+#  * Redistributions of source code must retain the above copyright
+#    notice, this list of conditions and the following disclaimer.
+#  * Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in the
+#    documentation and/or other materials provided with the distribution.
+#  * Neither the name of NVIDIA CORPORATION nor the names of its
+#    contributors may be used to endorse or promote products derived
+#    from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from typing import Dict, List, Optional
+
+from rich.console import Console
+from rich.table import Table
+
+
+class TelemetryDataExporterUtil:
+
+    TELEMETRY_DYNAMIC_METRICS_HEADER = [
+        "Metric",
+        "GPU",
+        "avg",
+        "min",
+        "max",
+        "p99",
+        "p95",
+        "p90",
+        "p75",
+        "p50",
+        "p25",
+    ]
+
+    TELEMETRY_CONSTANT_METRICS_HEADER = [
+        "Metric",
+        "GPU",
+        "Value",
+    ]
+
+    TELEMETRY_CONSTANT_METRICS = ["gpu_power_limit", "total_gpu_memory"]
+
+    TELEMETRY_GROUPS = {
+        "Power": ["gpu_power_usage", "gpu_power_limit", "energy_consumption"],
+        "Memory": ["gpu_memory_used", "total_gpu_memory"],
+        "Utilization": ["gpu_utilization"],
+    }
+
+    @staticmethod
+    def merge_telemetry_stats_json(
+        telemetry_stats: Optional[Dict], stats_and_args: Dict
+    ) -> None:
+        if telemetry_stats is not None:
+            stats_and_args.update({"telemetry_stats": telemetry_stats})
+
+    @staticmethod
+    def export_telemetry_stats_csv(telemetry_stats: Optional[Dict], csv_writer) -> None:
```

Comment:
Can you add the type hint for csv_writer? I know we don't do it elsewhere yet, but it would be good to start doing.

e.g.
```
from csv import writer as CsvWriter
def export_telemetry_stats_csv(telemetry_stats: Optional[Dict], csv_writer: CsvWriter) -> None:
```

Reply:
It's of type csv._writer. I think since it's an internal class, and the reason why we don't do it in other places. 
Should I add csv._writer?

Reply:
This doesn't work. It gives an error. 
It's of type csv._writer and do we want to use it? Asking this because this seems like a private method.

Reply:
Oh, that's strange. What's the error?

Reply:
Function "_csv.writer" is not valid as a typeMypy[valid-type](https://mypy.readthedocs.io/en/latest/_refs.html#code-valid-type)
Perhaps you need "Callable[...]" or a callback protocol?Mypy

Reply:
I tried a couple
```
import csv
csv_writer: csv.writer
``` gives MyPy error
```
import csv
csv_writer: csv._writer
``` gives error in the console
```
import _csv
csv_writer: _csv._writer
```
gives error in the console

Reply:
Looks like the csv writer object type is not accessable as it's defined in the C source code. But the following code seems to work with `mypy`:
```python
# foo.py
import csv
import _csv

with open("file.txt", "w") as f:
    w: _csv._writer = csv.writer(f)
```
and run 
```bash
mypy foo.py
```

ref: https://stackoverflow.com/questions/51264355/how-to-type-annotate-object-returned-by-csv-writer

Reply:
In type annotation, I'm unable to use _csv._writer or csv._writer or csv.writer.

Reply:
that's strange.. seems like this is a bit rare case so I'm fine dealing with the proper type hint later, but will defer to @dyastremsky's opinion on this.